patches/server/0814-Overridable-YAML-configuration.patch

```diff
@@ -0,0 +1,330 @@
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/configuration/OverrideSource.java
+@@ -0,0 +1,27 @@
++package io.papermc.paper.configuration;
++
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public interface OverrideSource {
++    /**
++     * An array of default override sources.
++     *
++     * @see EnvironmentOverrideSource
++     * @see PropertyOverrideSource
++     */
++    @NotNull
++    OverrideSource[] SOURCES = new OverrideSource[]{EnvironmentOverrideSource.INSTANCE, PropertyOverrideSource.INSTANCE};
++
++    @Nullable String getOverride(@NotNull String path, @NotNull String group);
++
++    static @Nullable String getOverrideDefault(@NotNull String path, @NotNull String group) {
++        for (OverrideSource source : SOURCES) {
++            String override = source.getOverride(path, group);
++            if (override != null) {
++                return override;
++            }
++        }
++        return null;
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/configuration/PropertyOverrideSource.java b/src/main/java/io/papermc/paper/configuration/PropertyOverrideSource.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..26f79ce75d4119d0484470f281c764bd2916b965
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/configuration/PropertyOverrideSource.java
+@@ -0,0 +1,27 @@
++package io.papermc.paper.configuration;
++
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Locale;
++
++/**
++ * An implementation of {@link OverrideSource}, which fetches values from JVM properties.
++ * The requested path will be transformed to lower case characters.
++ * <p>
++ * The group name will be prepended, followed by a dot (.).
++ */
++public final class PropertyOverrideSource implements OverrideSource {
++    @NotNull
++    public static final PropertyOverrideSource INSTANCE = new PropertyOverrideSource();
++
++    private PropertyOverrideSource() {
++
++    }
++
++    @Override
++    public @Nullable String getOverride(@NotNull String path, @NotNull String group) {
++        String key = group + "." + path;
++        return System.getProperty(key.toLowerCase(Locale.ROOT));
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+index 0544ac93513d3a274bfb53bb6120bd598f4d603b..c3e46500e2708fb79aabf7061f3a1ad1406a99ee 100644
+--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
++++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+@@ -144,4 +144,16 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
+ 
+         return this.worldGenSettings;
+     }
++
++    // Paper start
++    @org.jetbrains.annotations.Nullable
++    @Override
++    public String getStringRaw(String key) {
```

Comment:
Is this actually the best place to implement this ?
The parent of the `DedicatedServerProperties` class (`Settings`) already has the method `Settings#getOverride(String,String);String` that is called by the `Settings#getStringRaw` method.

Reply:
`getOverride` is private in Settings. Currently, the only subclass of it is `DedicatedServerProperties`. However, we don't know whether something else is going to subclass it in the future.

Reply:
Hmmm yes that is problematic tho with this implementation, environment vars would override even the command line flags respected by Settings#getStringRaw and the getOverride method respectively, wouldn't they ?

Reply:
Good point, however I think having environment variables override command line flags is a good idea. Let's say you have the command line flags specified in Dockerfile and you would like to override it at runtime.

Reply:
I think we will need more eyes on this. I'd personally expect command line flags to basically take priority over anything else.
Let's hope we get some maintainer/developer input 😅

Reply:
I would expect CLI > system prop > env var > file > default as a order, cause that's what the systems I encountered before do (for example, spring)
For your docker case, you can still make it work by having your flags as an env var in the start script and then you can override that.

Reply:
With the `Settings#getOverride(String,String)` method public, wouldn't it now be feasible to implement this behaviour in an override of the `Settings#getOverride(String,String)` method in the `DedicatedServerProperties` class ?
Seems like it might save you the hassle of manually fetching the property (and might just in general be a better place for the logic).

Tho this is really just a comment, this (at least from my end) is also a valid solution.

Reply:
I'm honestly not sure, Settings is an abstract class and we don't know whether something else will subclass it later. I personally wouldn't make another constructor or field there.

Reply:
No no, like override the 'getOverride' method in the dedicated server properties class. Sorry if I wasn't clear on that.

Reply:
Ah, although that would require transforming the `options` field afaik.

Reply:
You are right sir, I stand corrected. For some reason my brain still believed the `Settings#getOverride(String,String)` would just return null and we could easily check :+1: